### ISSUE TITLE
JWT Still valid after refresh

### ISSUE THREAD

**Issue by Echoju** [markitosgv/JWTRefreshTokenBundle#327]
Hello,

I ask for a JWT with a regular login by sending login&pwd by json.
Then I receive a JWT(1) & Refresh Token(1).

If I use the refresh Token(1) I received, so then I receive a new couple : JWT(2) & Refresh Token(2).

If I try to use JWT(1) and JWT(2) they both are valid.

Is this the expected behaviour ?
Is there a way so that there would be only one couple JWT+Refresh valid at the same time ?

Also is this possible to associated refresh token with an IP Address / Device so that no one else could use it ?

Thanks !

**Comment by mbabker** [comment]
> If I try to use JWT(1) and JWT(2) they both are valid.
> 
> Is this the expected behaviour ?

Yes.  This bundle on its own doesn't invalidate a JWT, you'd need to hook into the `gesdinet.refresh_token` event from this bundle or one of the authentication events from Symfony's security system to handle invalidating the older token.

> Also is this possible to associated refresh token with an IP Address / Device so that no one else could use it ?

You could extend the model in your own application to add this data.  You'd also need to handle checking it on your own.

**Comment by shakaran** [comment]
@Echoju, @mbabker's answer was right in 2022, but half of it has changed since, so it is worth updating rather than just closing.

**The old JWT staying valid is still expected**, and always will be. A JWT is verified by its signature and its expiry, with nothing consulted at verification time, which is the whole point of it — and the reason its lifetime is short and this bundle exists.

**What has changed is that it can now be withdrawn.** Lexik 3.x ships a token blocklist:

```yaml
lexik_jwt_authentication:
    blocklist_token:
        enabled: true
        cache: cache.app
```

That gives you `Lexik\Bundle\JWTAuthenticationBundle\Services\BlockedTokenManagerInterface`, and Lexik already blocks tokens on logout and on login failure. So "only one JWT valid at a time" is reachable in a way it was not when you asked.

Doing it on refresh is not something the bundle can do for you yet, for a reason worth knowing if you build it: a refresh request carries the refresh token, and not necessarily the old JWT, so there may be nothing to block. A listener of your own on `gesdinet.refresh_token` that reads the JWT from the current request when it is there, and blocks it, is the way today. I have noted it as something for the bundle to do when first-class Lexik 3.x support lands, since wiring it properly needs the request in `RefreshEvent`, which that event does not carry.

**On binding a token to an IP or device**, @mbabker's answer stands: extend the model, store what you want on it, and check it yourself. Two things are easier now than in 2022:

- `refresh_token_class` has always let you use your own model, and [mapping it yourself](https://github.com/markitosgv/JWTRefreshTokenBundle#mapping-the-entity-yourself) is now documented, which is what adding columns needs.
- `max_tokens_per_user` bounds how many sessions a user has at once, which is a blunter version of the same instinct and may be enough on its own.

Worth saying plainly, though: an IP check tends to sign out real users on mobile networks more often than it stops an attacker, and a device fingerprint sent by the client is only as trustworthy as the client. If the concern is a stolen refresh token, `single_use: true` does more for you — the token is replaced on every refresh, so a stolen one stops working the moment the real user refreshes.

Closing this. Thanks for asking it, and to @mbabker for answering it at the time.
